go-controller/pkg/clustermanager/userdefinednetwork/template/net-attach-def-template.go

```diff
@@ -1,18 +1,134 @@
-func (t NetAttachDefTemplate) RenderNetAttachDefManifest(_ *userdefinednetworkv1.UserDefinedNetwork) (*netv1.NetworkAttachmentDefinition, error) {
-	// TODO: implement NAD creation from UDN spec
-	return nil, errors.New("implement me")
+func RenderNetAttachDefManifest(udn *userdefinednetworkv1.UserDefinedNetwork) (*netv1.NetworkAttachmentDefinition, error) {
+	if udn == nil {
+		return nil, nil
+	}
+
+	cniNetConf, err := renderCNINetworkConfig(udn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to render CNI network config: %w", err)
+	}
+	cniNetConfRaw, err := json.Marshal(cniNetConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &netv1.NetworkAttachmentDefinition{
+		TypeMeta: newNetAttachDefTypeMeta(),
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            udn.Name,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(udn, udnKind)},
+			Labels: map[string]string{
+				labelUserDefinedNetwork: "",
+			},
+		},
+		Spec: netv1.NetworkAttachmentDefinitionSpec{
+			Config: string(cniNetConfRaw),
+		},
+	}, nil
+}
+
+func newNetAttachDefTypeMeta() metav1.TypeMeta {
+	const netAttachDefKind = "NetworkAttachmentDefinition"
+	var netAttachDefAPIVer = netv1.SchemeGroupVersion.String()
+	return metav1.TypeMeta{
+		APIVersion: netAttachDefAPIVer,
+		Kind:       netAttachDefKind,
+	}
+}
+
+func renderCNINetworkConfig(udn *userdefinednetworkv1.UserDefinedNetwork) (map[string]interface{}, error) {
+	networkName := udn.Namespace + "." + udn.Name
+	nadNamespaceName := util.GetNADName(udn.Namespace, udn.Name)
+
+	netConf := map[string]interface{}{
+		"cniVersion":       cniVersion,
+		"type":             ovnK8sCniOverlay,
+		"name":             networkName,
+		"netAttachDefName": nadNamespaceName,
+		"topology":         strings.ToLower(string(udn.Spec.Topology)),
+		"role":             strings.ToLower(string(udn.Spec.Role)),
+	}
+
+	if udn.Spec.MTU > 0 {
+		netConf["mtu"] = udn.Spec.MTU
+	}
+
+	primaryNetwork := udn.Spec.Role == userdefinednetworkv1.NetworkRolePrimary
+
+	if primaryNetwork && udn.Spec.Topology == userdefinednetworkv1.NetworkTopologyLocalnet {
+		return nil, fmt.Errorf("locannet topology is not supported for primary network")
+	}
+
+	if primaryNetwork {
+		const joinSubnetKey = "joinSubnets"
+		if len(udn.Spec.JoinSubnets) == 0 {
+			netConf[joinSubnetKey] = types.UserDefinedPrimaryNetworkJoinSubnetV4 + "," + types.UserDefinedPrimaryNetworkJoinSubnetV6
+		} else {
+			if err := validateSubnets(udn.Spec.JoinSubnets); err != nil {
+				return nil, fmt.Errorf("invalid join subnets: %w", err)
+			}
+			netConf[joinSubnetKey] = strings.Join(udn.Spec.JoinSubnets, ",")
+		}
+	}
+
+	if udn.Spec.IPAMLifecycle == userdefinednetworkv1.IPAMLifecyclePersistent {
+		if udn.Spec.Topology == userdefinednetworkv1.NetworkTopologyLayer2 ||
+			udn.Spec.Topology == userdefinednetworkv1.NetworkTopologyLocalnet {
+			netConf["allowPersistentIPs"] = strconv.FormatBool(true)
```

Comment:
I think you don't need this conversion, you will see unit test failure if you start using `ParseNADInfo` to validate the config

Reply:
best reference i could find is this https://github.com/kubevirt/community/pull/279/files#diff-cf2b59cd4b5f8edbed2374fdb9cf6f50623733ef278ad339ecde55ed7e97c3a6R466